### PR TITLE
qa: test runners home task

### PR DESCRIPTION
**Pull request by vladyslav-ko**

### CODE REVIEW

**Review by sforktesting — DISMISSED** [review]

File: src/main/java/com/reportportal/qa/pages/LaunchesPage.java [hunk 1]
```diff
@@ -2,17 +2,62 @@
...
+                    elementId)));
+
+        return getCommonMethods().getText(totalAmountOfTestsInBlock);
+    }
+
+    public String getPassedAmountOfTestsInBlock(String elementId) {
+        WebElement passedAmountOfTestsInBlock = getDriver().findElement(
+            By.xpath(
+                String.format(
+                    "(//*[@data-id='%s']//*[@class='executionStatistics__execution-statistics--1iek7']/a)[2]",
```
Comment (vikatyshchenko) [inline src/main/java/com/reportportal/qa/pages/LaunchesPage.java:35]: Why not store locators as variables to reuse them?
Or use @FindBy annotations?

File: src/main/java/com/reportportal/qa/steps/LaunchesSteps.java [hunk 2]
```diff
@@ -1,13 +1,31 @@
...
 
-    public void selectLaunchById(String launchId) {
+    public void goToLaunches() {
         launchesPage.launchesButton();
-        launchesPage.launchBlock(launchId);
+    }
+
+    public void selectLaunchById(String launchId) {
+        goToLaunches();
+        launchesPage.selectLaunchBlockById(launchId);
```
Comment (vikatyshchenko) [inline src/main/java/com/reportportal/qa/steps/LaunchesSteps.java:16]: Wouldn't it be better for methods to return an instance of a Steps class to provide less code duplication and make tests more readable?

**Review by vikatyshchenko — CHANGES_REQUESTED** [review]

File: test-output/Default Suite/taf-rp-launches.html
[the remark this answers is not in the document]
Comment (vladyslav-ko) [reply test-output/Default Suite/taf-rp-launches.html:1]: It was generated automatically. I deleted this package from branch and added this package to gitignore.

File: src/main/java/com/reportportal/qa/pages/LaunchesPage.java (hunk 1 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/pages/LaunchesPage.java:35]: I will do this in BDD home work.

File: src/main/java/com/reportportal/qa/steps/LaunchesSteps.java (hunk 2 above)
Comment (dkatsechau) [reply src/main/java/com/reportportal/qa/steps/LaunchesSteps.java:16]: However, in scope of this task this comment is valid.
Each PR is atomic and can not refer to the further modules.

File: src/main/java/com/reportportal/qa/pages/LaunchesPage.java (hunk 1 above)
Comment (dkatsechau) [reply src/main/java/com/reportportal/qa/pages/LaunchesPage.java:35]: Please, do not refer to the further modules. This PR is not about BDD.

**Review by sforktesting — CHANGES_REQUESTED** [review]

File: src/test/java/tests/junit/BaseJunit.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import com.reportportal.qa.testbase.CommonMethods;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+
+public class BaseJunit {
+
+    static LoginSteps loginSteps = new LoginSteps();
```
[the remark this answers is not in the document]
Comment (vladyslav-ko) [reply src/test/java/tests/junit/BaseJunit.java:16]: I fully agree with you. I will delete it.

**Review by sforktesting — APPROVED** [review]

File: test-output/Default Suite/taf-rp-launches.xml
[the remark this answers is not in the document]
Comment (vladyslav-ko) [reply test-output/Default Suite/taf-rp-launches.xml:1]: It was generated automatically. I deleted this package from branch and added this package to gitignore.

File: src/main/java/com/reportportal/qa/pages/LaunchesPage.java (hunk 1 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/pages/LaunchesPage.java]: Xpathes were moved into variables

File: src/main/java/com/reportportal/qa/steps/LaunchesSteps.java (hunk 2 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/steps/LaunchesSteps.java:16]: For me it is more readable when I write method after method, like:
launchSteps.clickOnButton();
launchSteps.selectElement(id);
launchSteps.selectAnotherElement(number);
Than it wil be like - launchSteps.clickOnButton().selectElement(id).selectAnotherElement(number);

**Review by vikatyshchenko — APPROVED** [review]

File: src/main/java/com/reportportal/qa/pages/LaunchesPage.java [hunk 4]
```diff
@@ -7,12 +7,39 @@
 
 public class LaunchesPage extends BasePage {
 
-    public void launchBlock(String elementId) {
+    public static final String LAUNCH_BLOCK = "(//*[@class='gridRow__grid-row-wrapper--1dI9K' and @data-id='%s']//*[@class='itemInfo__name-link--1ItPc'])[1]";
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/pages/LaunchesPage.java:10]: All this constants can be private.
Here and in other places.

File: src/test/java/tests/testng/SmokeTestNgTests3.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    void checkNameAndStartForLaunchesById(String launchId, String name, String start) {
+        launchesSteps.goToLaunches();
+        Assertions.assertEquals(launchesSteps.getLaunchNameText(launchId), name);
+        Assertions.assertEquals(launchesSteps.getLaunchStartText(launchId), start);
+    }
+
+    @DataProvider(name = "testDataCheckNameAndStartForLaunchesById")
+    public Object[][] testDataCheckNameAndStartForLaunchesById() {
+        return new Object[][]{
+            {"5", "Demo Api Tests", "29 days ago"},
```
Comment (dkatsechau) [inline src/test/java/tests/testng/SmokeTestNgTests3.java:22]: Is it still working with constant ```29 days ago```?
Are all tests passing without issues?

File: src/test/java/tests/testng/SmokeTestNgTests2.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    @DataProvider(name = "testDataCheckLaunchIsSuccessful")
+    public Object[][] testDataCheckLaunchIsSuccessful() {
+        return new Object[][]{
+            {"5"},
+            {"4"},
+            {"3"},
+        };
+    }
+
```
Comment (dkatsechau) [inline src/test/java/tests/testng/SmokeTestNgTests2.java:26]: Minor: Why again different format for empty lines?
Here and in other places.

File: src/test/java/tests/junit/SmokeJunitTests.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+@Execution(ExecutionMode.CONCURRENT)
+public class SmokeJunitTests extends BaseJunit {
+
+    @ParameterizedTest
+    @CsvSource({
+        "5, 1, Suite with retries, Here could be very important information about test-cases that are inside.",
+        "5, 2, Suite with nested steps, This is a suite level. Here you can handle the aggregated information per suite.",
+    })
+    @ExtendWith(ReportPortalExtension.class)
```
Comment (dkatsechau) [inline src/test/java/tests/junit/SmokeJunitTests.java:22]: Why use this annotation on methods, not on types?
Is it meant for something?
Here and in other places.

File: src/test/java/tests/testng/BaseTestNg.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+public class BaseTestNg {
+
+    LoginSteps loginSteps = new LoginSteps();
+    LaunchesSteps launchesSteps = new LaunchesSteps();
+    Logger log = LogManager.getLogger(CommonMethods.class);
+    SuiteSteps suiteSteps = new SuiteSteps();
+
+    @BeforeClass
+    public void openAndLogin() throws InterruptedException {
```
Comment (dkatsechau) [inline src/test/java/tests/testng/BaseTestNg.java:23]: Why this exception is throwing?

File: src/test/java/tests/junit/BaseJunit.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+
+public class BaseJunit {
+
+    LoginSteps loginSteps = new LoginSteps();
+    LaunchesSteps launchesSteps = new LaunchesSteps();
+    SuiteSteps suiteSteps = new SuiteSteps();
+
+    Logger log = LogManager.getLogger(CommonMethods.class);
```
Comment (dkatsechau) [inline src/test/java/tests/junit/BaseJunit.java:21]: Why does Logger for BaseJinit class use CommonMethods class?
Here and in other places.

File: src/test/java/tests/testng/BaseTestNg.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        getDriver();
```
Comment (dkatsechau) [inline src/test/java/tests/testng/BaseTestNg.java:24]: Why do you init driver for the entire test class?
Are your test methods linked inside the test class?

File: src/main/java/com/reportportal/qa/testbase/CommonMethods.java [hunk 11]
```diff
@@ -44,12 +49,23 @@ public static void setText(WebElement element, String text) {
...
+        wait.until(ExpectedConditions.visibilityOf(element)).click();
     }
 
     public void waitForElement(WebElement element) {
         log.info("Wait for element " + element);
         WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(WAITING_TIME));
         wait.until(ExpectedConditions.elementToBeClickable(element));
     }
+
+    public List<String> getElementTexts(String xpath) {
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/testbase/CommonMethods.java:61]: Method is not used anywhere.

File: src/main/java/com/reportportal/qa/testbase/CommonMethods.java [hunk 12]
```diff
@@ -44,12 +49,23 @@ public static void setText(WebElement element, String text) {
...
+        log.info("List of elements by xpath " + xpath);
+        List<WebElement> elements = getDriver().findElements(By.xpath(xpath));
+        List<String> elementTexts = new ArrayList<>();
+        for (WebElement element : elements) {
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/testbase/CommonMethods.java:65]: Why not to use stream?

File: src/main/java/com/reportportal/qa/steps/LaunchesSteps.java [hunk 13]
```diff
@@ -1,13 +1,31 @@
 package com.reportportal.qa.steps;
 
 import com.reportportal.qa.pages.LaunchesPage;
+import java.util.List;
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/steps/LaunchesSteps.java:4]: Not used import.

File: src/main/java/com/reportportal/qa/pages/LaunchesPage.java [hunk 14]
```diff
@@ -7,12 +7,39 @@
...
-                String.format("//*[@class='gridRow__grid-row-wrapper--1dI9K' and @data-id='%s']",
-                    elementId)));
+            By.xpath(String.format(LAUNCH_BLOCK, elementId)));
         getCommonMethods().waitForElementAndClick(launchesBlockButton);
+    }
+
+    public String getTotalAmountOfTestsInBlock(String elementId) {
+        WebElement totalAmountOfTestsInBlock = getDriver().findElement(
+            By.xpath(String.format(TOTAL_AMOUNT_OF_TESTS, elementId)));
+        return getCommonMethods().getText(totalAmountOfTestsInBlock);
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/pages/LaunchesPage.java:25]: Why access static method from instance reference?
Here and in other places.

File: src/test/java/tests/testng/BaseTestNg.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+        log.info("Navigate to login page");
+        loginSteps.loginDefaultUser();
+        Assertions.assertTrue(loginSteps.checkThatWeLoggedInSuccessfully());
+    }
+
+    @AfterClass
+    public void navigateToLaunchesAndCloseBrowser() {
```
Comment (dkatsechau) [inline src/test/java/tests/testng/BaseTestNg.java:31]: Why do this method have such name?

File: src/test/java/tests/testng/SelectLaunchByIdAndCheckSuiteDescriptionTestNgTest.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    @ExtendWith(ReportPortalExtension.class)
+    void selectLaunchByIdAndCheckSuiteDescription(String launchId, String suiteId, String suiteName,
+        String suiteDescription) {
+        launchesSteps.selectLaunchById(launchId);
+        Assertions.assertEquals(suiteSteps.getSuiteNameText(suiteId), suiteName);
+        Assertions.assertEquals(suiteSteps.getSuiteDescriptionText(suiteId), suiteDescription);
+    }
+
+    @DataProvider(name = "selectLaunchByIdAndCheckSuiteDescription")
+    public Object[][] testDataSelectLaunchById() {
```
Comment (dkatsechau) [inline src/test/java/tests/testng/SelectLaunchByIdAndCheckSuiteDescriptionTestNgTest.java:21]: Your dataprovider, befoclass and afterclass methods are public,
whyle test method are package-private
Is it for purpose?
Here and in other places.

File: src/main/java/com/reportportal/qa/steps/SuiteSteps.java [hunk 17]
```diff
@@ -0,0 +1,16 @@
+package com.reportportal.qa.steps;
+
+import com.reportportal.qa.pages.SuitePage;
+
+public class SuiteSteps {
+
+    SuitePage suitePage = new SuitePage();
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/steps/SuiteSteps.java:7]: Why not private?

File: src/main/java/com/reportportal/qa/steps/LaunchesSteps.java [hunk 18]
```diff
@@ -1,13 +1,31 @@
...
 
 public class LaunchesSteps {
 
     LaunchesPage launchesPage = new LaunchesPage();
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/steps/LaunchesSteps.java:8]: Why not private?
Here and in other places.

File: src/main/java/com/reportportal/qa/pages/SuitePage.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+import org.openqa.selenium.WebElement;
+
+public class SuitePage extends BasePage {
+
+    public static final String SUITE_NAME = "((//*[@class='gridRow__grid-row-wrapper--1dI9K'])[%s]//div[@class='tooltip__tooltip-trigger--3Z4Hc itemInfo__name--27fwI'])[1]";
+    public static final String SUITE_DESCRIPTION = "((//*[@class='gridRow__grid-row-wrapper--1dI9K'])[%s]//div[@class='markdownViewer__markdown-viewer--13oxq mode-default'])[2]";
+
+    public String getSuiteNameText(String elementId) {
+        WebElement suiteName = getDriver().findElement(
+            By.xpath(String.format(SUITE_NAME,elementId)));
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/pages/SuitePage.java:15]: Why not formatted? Space should be placed after comma.
Here and in other places.

File: src/main/java/com/reportportal/qa/pages/BasePage.java [hunk 20]
```diff
@@ -8,16 +8,17 @@
...
 public class BasePage {
 
-    private CommonMethods commonMethods = new CommonMethods();;
+    private final CommonMethods commonMethods = new CommonMethods();
 
     public CommonMethods getCommonMethods() {
         return commonMethods;
     }
 
+    public static final String LAUNCHES_BUTTON = "//div[@class='sidebarButton__sidebar-nav-btn--1prEO']/a[contains(@href,'launches')]";
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/pages/BasePage.java:17]: What was happened with file organization?
Code Conventions for the Java: https://www.oracle.com/java/technologies/javase/codeconventions-fileorganization.html

File: src/test/java/tests/junit/SmokeJunitTests.java (hunk 7 above)
Comment (vladyslav-ko) [reply src/test/java/tests/junit/SmokeJunitTests.java:22]: @ExtendWith(ReportPortalExtension.class): This annotation is from the JUnit Jupiter framework and is used to register the ReportPortalExtension class as an extension for the test class. It allows the integration of the ReportPortal reporting tool with the test execution.

@Execution(ExecutionMode.CONCURRENT): This annotation is used to specify the execution mode of the test class. In this case, the value is set to ExecutionMode.CONCURRENT, which means that the test methods in this class can be executed concurrently.

@ParameterizedTest: This annotation is from the JUnit Jupiter framework and is used to indicate that the annotated method is a parameterized test. Parameterized tests allow you to run the same test method multiple times with different arguments.

@CsvSource: This annotation is used to provide comma-separated values (CSV) as a source for the parameterized test. Each row in the CSV represents a set of arguments for a test invocation.

@ValueSource: This annotation is used to provide a single-dimensional array of values as a source for the parameterized test. Each value in the array represents an argument for a test invocation.

File: src/test/java/tests/testng/BaseTestNg.java (hunk 8 above)
Comment (vladyslav-ko) [reply src/test/java/tests/testng/BaseTestNg.java:23]: I have no idea)
I will delete this

File: src/test/java/tests/junit/BaseJunit.java (hunk 9 above)
Comment (vladyslav-ko) [reply src/test/java/tests/junit/BaseJunit.java:21]: I will fix it

File: src/main/java/com/reportportal/qa/testbase/CommonMethods.java (hunk 11 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/testbase/CommonMethods.java:61]: I will delete it

File: src/main/java/com/reportportal/qa/testbase/CommonMethods.java (hunk 12 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/testbase/CommonMethods.java:65]: According to the previous comment, I will delete this method.

File: src/main/java/com/reportportal/qa/steps/LaunchesSteps.java (hunk 13 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/steps/LaunchesSteps.java:4]: I will delete it

File: src/test/java/tests/testng/BaseTestNg.java (hunk 15 above)
Comment (vladyslav-ko) [reply src/test/java/tests/testng/BaseTestNg.java:31]: i will rename it - closeBrowser

File: src/test/java/tests/testng/SelectLaunchByIdAndCheckSuiteDescriptionTestNgTest.java (hunk 16 above)
Comment (vladyslav-ko) [reply src/test/java/tests/testng/SelectLaunchByIdAndCheckSuiteDescriptionTestNgTest.java:21]: It doesn't have any purpose. All test methods now are public.

File: src/main/java/com/reportportal/qa/pages/SuitePage.java (hunk 19 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/pages/SuitePage.java]: now it is formatted

File: src/main/java/com/reportportal/qa/pages/BasePage.java (hunk 20 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/pages/BasePage.java]: I apologize for this. Now we have:
- private static final
- private final

File: src/test/java/tests/testng/SmokeTestNgTests3.java [hunk 21]
```diff
@@ -0,0 +1,27 @@
+package tests.testng;
+
+import com.epam.reportportal.junit5.ReportPortalExtension;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+public class SmokeTestNgTests3 extends BaseTestNg{
```
[the remark this answers is not in the document]
Comment (vladyslav-ko) [reply src/test/java/tests/testng/SmokeTestNgTests3.java]: fixed

File: src/main/java/com/reportportal/qa/pages/BasePage.java [hunk 22]
```diff
@@ -8,16 +8,17 @@
...
+
     public void launchesButton() {
```
[the remark this answers is not in the document]
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/pages/BasePage.java]: fixed

File: src/main/java/com/reportportal/qa/pages/LaunchesPage.java (hunk 4 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/pages/LaunchesPage.java]: fixed

File: src/test/java/tests/testng/SmokeTestNgTests3.java (hunk 5 above)
Comment (vladyslav-ko) [reply src/test/java/tests/testng/SmokeTestNgTests3.java:22]: When i created this PR - all tests were green. Now I have some problems with docker, when i resolve them, I will change data for tests and attach the screenshot with green run.

File: src/test/java/tests/testng/SmokeTestNgTests2.java (hunk 6 above)
Comment (vladyslav-ko) [reply src/test/java/tests/testng/SmokeTestNgTests2.java]: Now everything in one code style.

File: src/main/java/com/reportportal/qa/steps/SuiteSteps.java (hunk 17 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/steps/SuiteSteps.java]: private now

File: src/main/java/com/reportportal/qa/steps/LaunchesSteps.java (hunk 18 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/steps/LaunchesSteps.java]: private now

File: src/test/java/tests/junit/SmokeJunitTests.java (hunk 7 above)
Comment (dkatsechau) [reply src/test/java/tests/junit/SmokeJunitTests.java:20]: I only asked about one annotation, that I was commented.
Did you use it on types? Does it work when used on types?

File: src/test/java/tests/testng/BaseTestNg.java (hunk 10 above)
Comment (dkatsechau) [reply src/test/java/tests/testng/BaseTestNg.java:23]: Somehow you create new driver in Junit```@BeforeEach``` annotated method - and it's ok.
Please, don't create driver for class, unless your tests are binded with consecutive logic.

File: src/main/java/com/reportportal/qa/pages/LaunchesPage.java (hunk 14 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/pages/LaunchesPage.java]: fixed

File: src/test/java/tests/junit/SmokeJunitTests.java (hunk 7 above)
Comment (vladyslav-ko) [reply src/test/java/tests/junit/SmokeJunitTests.java:20]: Move this annotation to class
image

File: src/test/java/tests/testng/BaseTestNg.java (hunk 10 above)
Comment (vladyslav-ko) [reply src/test/java/tests/testng/BaseTestNg.java:23]: image

File: pom.xml [hunk 23]
```diff
@@ -80,6 +80,48 @@
...
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-params</artifactId>
+      <version>5.9.2</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.cucumber</groupId>
```
Comment (dkatsechau) [inline pom.xml:96]: Do we cucumber dependencies this now?

**Pull request merged**